### ISSUE TITLE
DNS lookup for a headless service with `publishNotReadyAddresses: false` returns non-ready pod IPs as well as ready pod IPs

### ISSUE THREAD

**Issue by evanskinner** [coredns/coredns#4594]
**What happened**:

Given the following outputs from kubectl (in k8s version: v1.19.10+IKS)

```
$ kubectl describe svc portal-4a01960c-www-ready
Name:              portal-4a01960c-www-ready
Namespace:         apic
Labels:            app.kubernetes.io/instance=portal
                   app.kubernetes.io/managed-by=ibm-apiconnect
                   app.kubernetes.io/name=www
                   com.ibm.apiconnect/multisiteha-mode=none
                   com.ibm.apiconnect/portal-compatibility=2
Annotations:       apiconnect-operator/hash: baf224cd455545d807d20823d86c370221ac80bd995e7e02573584eed2a960c9
                   productChargedContainers: all
                   productID: 2c717d4ecc924aa5ac345015b55492eb
                   productMetric: PROCESSOR_VALUE_UNIT
                   productName: IBM API Connect Enterprise
                   productVersion: 10.0.1.2-1987-eus
Selector:          app.kubernetes.io/instance=portal,app.kubernetes.io/managed-by=ibm-apiconnect,app.kubernetes.io/name=www
Type:              ClusterIP
IP:                None
Port:              director  3009/TCP
TargetPort:        3009/TCP
Endpoints:         172.30.127.169:3009
Port:              web  4443/TCP
TargetPort:        4443/TCP
Endpoints:         172.30.127.169:4443
Session Affinity:  None
Events:            <none>

$ kubectl describe endpoints portal-4a01960c-www-ready
Name:         portal-4a01960c-www-ready
Namespace:    apic
Labels:       app.kubernetes.io/instance=portal
              app.kubernetes.io/managed-by=ibm-apiconnect
              app.kubernetes.io/name=www
              com.ibm.apiconnect/multisiteha-mode=none
              com.ibm.apiconnect/portal-compatibility=2
              service.kubernetes.io/headless=
Annotations:  <none>
Subsets:
  Addresses:          172.30.127.169
  NotReadyAddresses:  172.30.127.170 172.30.127.171
  Ports:
    Name      Port  Protocol
    ----      ----  --------
    director  3009  TCP
    web       4443  TCP

Events:  <none>

$ kubectl get po -o wide | grep www-
portal-4a01960c-www-0                                             2/2     Running     0          9h      172.30.127.169   10.184.234.119   <none>           <none>
portal-4a01960c-www-1                                             0/2     Running     0          9h      172.30.127.170   10.184.234.67    <none>           <none>
portal-4a01960c-www-2                                             0/2     Running     0          9h      172.30.127.171   10.184.234.69    <none>           <none>
```

I would not then expect a DNS lookup of the service to return all pods as only one of them is ready:

```
$ k exec -ti portal-4a01960c-www-0 /bin/bash
Defaulting container name to admin.
Use 'kubectl describe pod/portal-4a01960c-www-0 -n apic' to see all of the containers in this pod.
bash-4.4$ nslookup -nosearch portal-4a01960c-www-ready.apic.svc.cluster.local
Server:		172.21.0.10
Address:	172.21.0.10#53

Name:	portal-4a01960c-www-ready.apic.svc.cluster.local
Address: 172.30.127.169
Name:	portal-4a01960c-www-ready.apic.svc.cluster.local
Address: 172.30.127.170
Name:	portal-4a01960c-www-ready.apic.svc.cluster.local
Address: 172.30.127.171
```

**What you expected to happen**:

I expected this to happen, only the ready pod is returned

```
$ k exec -ti portal-4a01960c-www-0 /bin/bash
Defaulting container name to admin.
Use 'kubectl describe pod/portal-4a01960c-www-0 -n apic' to see all of the containers in this pod.
bash-4.4$ nslookup -nosearch portal-4a01960c-www-ready.apic.svc.cluster.local
Server:		172.21.0.10
Address:	172.21.0.10#53

Name:	portal-4a01960c-www-ready.apic.svc.cluster.local
Address: 172.30.127.169
```

**How to reproduce it (as minimally and precisely as possible)**:

- Create a service like the `portal-4a01960c-www-ready` attached and have a statefulset (a deployment would likely also work) as the endpoints for the service.
- Ensure that 1 out of 3 of your pods are ready
- do an nslookup of the service and note that all 3 pod IPs are returned when it sihould only be 1

**Anything else we need to know?**:

This used to work with coredns 1.7.0, but somewhere after 1.7.0 it satrted returning ALL pods not jut the ready ones.

**Environment**:

- the version of CoreDNS: 1.8.3
- Corefile: no core.
- logs, if applicable: nothings interesting shows in logs that I can see
- OS (e.g: `cat /etc/os-release`): coredns kubernetes container
- Others: N/A

Core DNS config:

```
    # Add your CoreDNS customizations as import files.
    # Refer to https://cloud.ibm.com/docs/containers?topic=containers-cluster_dns for details.
    .:53 {
        errors
        health {
            lameduck 10s
        }
        ready
        kubernetes cluster.local in-addr.arpa ip6.arpa {
            pods insecure
            fallthrough in-addr.arpa ip6.arpa
            ttl 30
        }
        prometheus :9153
        forward . /etc/resolv.conf {
            max_concurrent 1000
        }
        cache 30
        loop
        reload
        loadbalance
    }
```

**Comment by evanskinner** [comment]
This might be totally wrong but is it possible the bug was introduced (in October 2020) by https://github.com/coredns/coredns/pull/4209 and fixed (3 days ago, so not in 1.8.3) by https://github.com/coredns/coredns/pull/4580 ?

**Comment by chrisohaver** [comment]
Yes, fixed in #4580.

**Comment by evanskinner** [comment]
Thank you @chrisohaver
